Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp

```diff
@@ -76,6 +110,107 @@ TextureMapperLayer::~TextureMapperLayer()
     removeFromParent();
 }
 
+void TextureMapperLayer::flattenDescendantLayersIfNecessary(TextureMapperPaintOptions& options)
+{
+    // Preprocess is called as a first step before painting any layers
+
+    // Layers are collected in depth first, post-order
+    Vector<TextureMapperLayer*> layersToFlatten;
+    collectLayersToFlattenRecursive(layersToFlatten);
+
+    for (auto* layer : layersToFlatten)
+        layer->flatten(options);
+}
+
+void TextureMapperLayer::collectLayersToFlattenRecursive(Vector<TextureMapperLayer*>& layers)
+{
+    // Traverse all children
+    for (auto* child : m_children) {
+        // Recursively collect from the child's descendants
+        child->collectLayersToFlattenRecursive(layers);
+
+        // If the child satisfies the condition, add it to the layers vector
+        if (child->flattensAsLeafOf3DSceneOr3DPerspective())
+            layers.append(child);
+    }
+}
+
+void TextureMapperLayer::flatten(TextureMapperPaintOptions& options)
+{
+    // Flattens descendant layers to into this planes z=0. In theory we could
+    // project each sublayer to this layer's plane but in practice that doesn't
+    // result in pixel perfect outcome and that's why each sublayer needs to be
+    // rendered into this layer's in 2D root.
+
+    // Reset transformations for flattening root 2D
+    SetForScope scopedLocalTransform(m_layerTransforms.localTransform, TransformationMatrix());
+    SetForScope scopedFutureLocalTransform(m_layerTransforms.futureLocalTransform, TransformationMatrix());
+    m_layerTransforms.combined = { };
+    m_layerTransforms.combinedForChildren = { };
+    m_layerTransforms.futureLocalTransform = { };
+    m_layerTransforms.futureCombined = { };
+    m_layerTransforms.futureCombinedForChildren = { };
+    ComputeTransformData data;
+    if (m_state.maskLayer)
+        m_state.maskLayer->computeTransformsRecursive(data);
+    if (m_state.replicaLayer)
+        m_state.replicaLayer->computeTransformsRecursive(data);
+    if (m_state.backdropLayer)
+        m_state.backdropLayer->computeTransformsRecursive(data);
+    for (auto* child : m_children)
+        child->computeTransformsRecursive(data);
```

Comment:
TextureMapperLayer::paint also calls computeTransformsRecursive for all layers.
It's redundant to compute transforms for a layer multiple times.
In my WIP patch, it is computed only once.

Reply:
For flattening 2D root, child transforms need to be calculated for that root. After flattening is done, then transforms need to be recalculated taking account flattened layer rect. This is relatively cheap operation because trees are usually not very deep and calculation involved is not complex

Reply:
I think this approach doesn't work with a backdrop filter.
A backdrop layer renders the root layer with identity matrix into a temporary buffer.
https://github.com/WebKit/WebKit/blob/a10ee6a9b1ee5949ced6cfdab1a393243e65a9b3/Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp#L725-L729
`computeTransformsRecursive` should be done before it.

Reply:
You right, this seems to fail. I'm working on a fix

Reply:
Backdrops are now working according to spec. I added new test case
https://bug-281079-attachments.webkit.org/attachment.cgi?id=473058